rados/read_op_exec.go

```diff
@@ -0,0 +1,114 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <stdlib.h>
+// #include <rados/librados.h>
+//
+import "C"
+
+import (
+	"runtime"
+	"unsafe"
+)
+
+// ReadOpExecStep - step for exec operation code.
+type ReadOpExecStep struct {
+	charPtr *C.char
+	size    C.size_t
+	prval   *C.int
+}
+
+// newExecStepOp - init new *execStepOp.
+func newReadOpExecStep() *ReadOpExecStep {
+	es := &ReadOpExecStep{
+		charPtr: nil,
+		size:    0,
+		prval:   (*C.int)(C.malloc(C.sizeof_int)),
+	}
+	*es.prval = 0
+
+	runtime.SetFinalizer(es, func(es *ReadOpExecStep) {
+		if es != nil {
+			es.free()
+			es.freeBuffer()
+			es = nil
+		}
+	})
+	return es
+}
+
+// free - will be called to free any resources, especially C memory.
+func (es *ReadOpExecStep) free() {
+	if es.prval != nil {
+		C.free(unsafe.Pointer(es.prval))
+		es.prval = nil
+	}
+}
+
+// freeBuffer - releases C allocated buffer. It is separated from es.free() because lifespan of C allocated buffer is
+// longer than lifespan of read operation.
+func (es *ReadOpExecStep) freeBuffer() {
+	if es.charPtr != nil {
+		C.free(unsafe.Pointer(es.charPtr))
+		es.charPtr = nil
+	}
+}
+
+// update - update state operation.
+func (es *ReadOpExecStep) update() error {
+	return getError(*es.prval)
+}
+
+// Bytes - go slice of bytes allocated in C.
+func (es *ReadOpExecStep) Bytes() []byte {
+	return C.GoBytes(unsafe.Pointer(es.charPtr), C.int(es.size))
+}
+
+// Exec - execute an OSD class method on an object See rados_exec() for general description.
+// The output buffer is allocated on the heap; the caller is expected to release that memory with
+// rados_buffer_free(). The buffer and length pointers can all be NULL, in which case they are not filled in.
```

Comment:
I'd remove this. This seems particularly tied to how this is meant to be used in C where the Go API you're adding avoids foisting this complexity on the user of the API.

Reply:
So we should add something like `ReadOp.operateCalled` bool flag? I don't see another way to ensure that ReadOp.Operate is called.